### PR TITLE
Make withRouter() update inside static containers

### LINKED ISSUE #3439: Make withRouter() update in response to context changes
I wonder if you’d be up for accepting a PR like #3430 but for `withRouter()` (#3352). This would eliminate the problem of “how do I connect to the router state in Redux in a deep component” because people would just wrap their deep components with `withRouter` and be guaranteed to get the right `params`.

If this makes sense to you, I’d submit a PR that moves the context fix to be used by `withRouter`, and change `Link` to use `withRouter` internally.

### PR DESCRIPTION
**Pull request by gaearon**
Fixes #3439.

### CODE REVIEW

File: modules/ContextUtils.js [hunk 1]
```diff
@@ -111,8 +112,12 @@ export function ContextSubscriber(name) {
       this[unsubscribeKey] = null
     },
 
-    [handleContextUpdateKey](eventIndex) {
+    [handleContextUpdateKey](eventIndex, nextValue) {
       if (eventIndex !== this.state[lastRenderedEventIndexKey]) {
+        if (this.context[name] !== nextValue) {
+          // React uses a stale value so we update it manually.
+          this.context[name] = nextValue;
```
Comment (gaearon) [inline modules/ContextUtils.js]: I’m not proud of this but it works fine. I’ll keep an eye on this not breaking in the React repo 😄

File: modules/ContextUtils.js [hunk 2]
```diff
@@ -43,8 +43,9 @@ export function ContextProvider(name) {
     },
 
     componentDidUpdate() {
+      const nextValue = this.getChildContext()[name]
```
Comment (gaearon) [inline modules/ContextUtils.js]: This is safe because React makes no guarantees about when this method can be called

**Comment by taion** [comment]
I think on net this is a bit messier than the earlier implementation that just updated the router object in-place.

**Comment by gaearon** [comment]
Yea. On the other hand, this keeps the desired property of `Context*` being generic. If they aren’t updating context to new objects, I’m not sure how much useful they would be as an outside module.

**Comment by taion** [comment]
I mean, it's an unfortunate caveat to have, but I'm guessing it's probably slightly more performant to not recreate the object every time. Having to keep the same object identity is a caveat that I can live with.

**Comment by gaearon** [comment]
>it's probably slightly more performant to not recreate the object every time

I think this would only be relevant if navigation was something that’s performed hundreds of time per second. I wouldn’t expect this to have any measurable performance in this particular scenario.

Happy to change it though. You sure?

**Comment by gaearon** [comment]
What I like about this is while it’s a net messier, the messy bits are all in the “polyfill”. As opposed to spreading through the legit code.

**Comment by gaearon** [comment]
I’m fairly content with what I posted before but I added 7b3abdb39e5a3c74eb114f1bc52a35daecf2cfce as a minimally invasive alternative approach. It adds `object-assign` as a dependency but React already uses it anyway.

**Comment by taion** [comment]
cc @jquense who's done some stuff to work around this issue. At this point I'm really torn.

**Comment by gaearon** [comment]
Ultimately we can go either way and switch if we find problems later. I’d probably use the last commit because it doesn’t touch `this.context` so if React (unlikely) adds a warning on setting to it, this wouldn’t affect us.

**Comment by taion** [comment]
Yeah, I'm fine with imposing a bit of extra work on the users of this polyfill I think. Probably I'd be happier mutating my context objects in place rather than worrying that setting `this.context` will break. It's all a workaround anyway.

**Comment by gaearon** [comment]
In case we want to change the mind later, the test case I added should catch any mistakes.

**Comment by taion** [comment]
I'm happy with this for now.

**Comment by gaearon** [comment]
@taion One last problem is this still doesn’t update containers with shallow equality check. We can either revert the last commit or add a bogus prop just to force an update.

**Comment by gaearon** [comment]
Alternatively, we can inject `params` and `location` as props to `useRouter()`, in addition to `router` itself. Then it wouldn’t matter.

**Comment by taion** [comment]
You mean in case someone does `withRouter(connect(whatever)(MyComponent))`?

**Comment by gaearon** [comment]
Yeah.

**Comment by taion** [comment]
I like the idea of `withRouter` injecting the additional props – makes `withRouter`-ed components work identically to route components, which seems like a nice symmetry.

**Pull request merged**